Add drop replica alter support

sundy-li opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

If the replicated table has 3 replicas, but one replica is crashed and never be back to normal again, we should enable to remove this replica by alter commands.


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `Alter table drop replica replica_name` support. This fixes #7080.

Detailed description / Documentation draft:
Already in this commit.
 .

sundy-li committed:
Add drop replica alter support

sundy-li committed:
fix merge state

Fix style && build

sundy-li committed:
Add tests and docs

Fix integration-tests

filimonov commented:
It should not be possible to remove active replica this way, only stale/inactive.

filimonov commented:
https://github.com/ClickHouse/ClickHouse/issues/7080

sundy-li committed:
only drop inactive replica

update doc

Increase timeout to release the zookeeper Ephemeral nodes

Fix code comment

use PartitionManager

make integrations test passed

qoega commented on src/Storages/StorageReplicatedMergeTree.h:
```diff
@@ -115,6 +115,10 @@ class StorageReplicatedMergeTree final : public ext::shared_ptr_helper<StorageRe
       */
     void drop() override;
 
+    /** Removes a specific replica from Zookeeper. If replica is local, it works same as `drop` method.
```
Other code and comments tell it is impossible to drop local replica. Incorrect comment?

sundy-li replied:
Yes. I will fix the comment, the local replica can only drop by `DROP TABLE`

qoega commented on docs/en/sql-reference/statements/alter.md:
````diff
@@ -197,6 +197,16 @@ Constraint check *will not be executed* on existing data if it was added.
+ALTER TABLE [db].name DROP REPLICA replica_name;
+```
+
+Queries will remove the replica path in zookeeper, it's useful when you want to decrease your replica factor. It will only drop the inactive replica, and it can't drop local replica, please use `DROP TABLE`.
````
What happens with data on that replica? It remains on remote server?

sundy-li replied:
Yes, it remains. We can't control that.  The main reason we call `DROP REPLICA` is that we should enable remove the 'dead' replica.

sundy-li commented:
BTW What's `Yandex synchronization check`?   I can't look into the Details.

I'll look into the `Integration tests`, seems the Ephemeral node `/is_acitve` not released when we kill the `node_1_2`

alexey-milovidov commented:
> What's Yandex synchronization check?

It's about a mirror of ClickHouse source code that is automatically created in bigger internal monorepository. You don't have to worry about this check.

alexey-milovidov commented:
For somewhat reason, newly added integration test does not pass.

tavplubix commented:
Maybe it should be implemented as `SYSTEM` query instead of `ALTER`?
Something like `SYSTEM DROP REPLICA 'replica_name' FROM table_name` to remove dead replica of existing table and `SYSTEM DROP REPLICA 'replica_name' FROM '/path/to/table/in/zk/'` to remove dead replica of some table, which does not exists on the local server.
The second query will be useful to remove garbage from ZooKeeper after the local replica was dropped, but metadata of dead replica are left.

sundy-li commented:
> Maybe it should be implemented as `SYSTEM` query instead of `ALTER`?
> Something like `SYSTEM DROP REPLICA 'replica_name' FROM table_name` to remove dead replica of existing table and `SYSTEM DROP REPLICA 'replica_name' FROM '/path/to/table/in/zk/'` to remove dead replica of some table, which does not exists on the local server.
> The second query will be useful to remove garbage from ZooKeeper after the local replica was dropped, but metadata of dead replica are left.

Good idea, I will consider it later. (Now I am just a little busy about work)

sundy-li committed:
add SYSTEM DROP REPLICA

amudong committed:
fix: can not drop local table and replace ALTER with SYSTEM

fix test_drop_replica

fix drop replica '/path/to/zk/' ending in '/' and update doc

amudong commented:
> Maybe it should be implemented as `SYSTEM` query instead of `ALTER`?
> Something like `SYSTEM DROP REPLICA 'replica_name' FROM table_name` to remove dead replica of existing table and `SYSTEM DROP REPLICA 'replica_name' FROM '/path/to/table/in/zk/'` to remove dead replica of some table, which does not exists on the local server.
> The second query will be useful to remove garbage from ZooKeeper after the local replica was dropped, but metadata of dead replica are left.

Yes, I have implemented the way to replace the `ALTER` query with  `SYSTEM`, and support the two calling methods you mentioned. I will commit it later.Also, we want to add some methods like `SYSTEM DROP REPLICA 'replica_name'`, `SYSTEM DROP REPLICA FROM database  'replica_name'` and `SYSTEM DROP REPLICA FROM database.table 'replica_name'`.

amudong committed:
add drop relica database and the whole replica

fix removeReplicaByZKPath

fix bug: add zkpath empty judge

fix: rewrite code

delete useless code.

fix:ast fromat

fix bug

add test_drop_replica

add drop_replica doc

add drop databse checkAccess

refactor dropReplica

update tests

add static method StorageReplicatedMergeTree::dropReplicaByZkPath

update doc and delete useless code

fix conflict

fix doc

fix doc

fix StorageReplicatedMergeTree::dropReplica

fix bug

delete useless code

tavplubix commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -400,6 +403,96 @@ void InterpreterSystemQuery::restartReplicas(Context & system_context)
+            {
+                storage_replicated->getStatus(status);
+
+                if (status.replica_path.compare(query.replica_zk_path + "/replicas/" + status.replica_name) == 0)
```
If local table with such `replica_path` exists and `DROP TABLE` should be used, then it's better to throw exception right here without call of `StorageReplicatedMergeTree::dropReplica(...)`

tavplubix commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -400,6 +403,96 @@ void InterpreterSystemQuery::restartReplicas(Context & system_context)
+    LOG_INFO(log, "Removing replica {}", remote_replica_path);
+    /// It may left some garbage if replica_path subtree are concurently modified
+    auto zookeeper = context.getZooKeeper();
+    zookeeper->tryRemoveRecursive(remote_replica_path);
```
We should check that replica is not active before removing it. Also we probably should remove it the same way as `StorageReplicatedMergeTree::dropReplica(...)` does. It can be done by extracting some part of `StorageReplicatedMergeTree::dropReplica(...)` to static method.

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -815,6 +715,113 @@ static time_t tryGetPartCreateTime(zkutil::ZooKeeperPtr & zookeeper, const Strin
+        throw Exception("Table was not dropped because ZooKeeper session has expired.", ErrorCodes::TABLE_WAS_NOT_DROPPED);
+
+    if (is_drop_table)
+        replica_is_active_node = nullptr;
```
Isn't it done in `StorageReplicatedMergeTree::shutdown()` which calls `ReplicatedMergeTreeRestartingThread::partialShutdown()`?

amudong committed:
rewrite static StorageReplicatedMergeTree::dropReplica

amudong committed:
format comment

amudong committed:
Optimize code

cw9 commented:
Thanks for adding this!
